virtool/api/utils.py

```diff
@@ -180,50 +183,77 @@ async def paginate_aggregate(
     'sort': field to sort by
     'skip_count': number of documents to skip
 
-    :param collection: the database collection
-    :param db_query: a query derived from user supplied - affects found count
-    :param url_query: the raw URL query; used to get the `page` and `page_count` values
-    :param sort: a field to sort by
+    :param collection: the database collection\
+    :param page: the page number to return (starts at one)
+    :param per_page: the number of items to return per page
+    :param client_query: a query derived from user supplied - affects found count
     :param base_query: a query always applied to the search
+    :param transforms: a list of transforms to apply to the query
+    :param projection: the projection to apply to the returned documents
+    :param sort: a field to sort by
     :param reverse: reverse the sort order
 
     """
-    try:
-        page = int(url_query["page"])
-    except (KeyError, ValueError):
-        page = 1
-
-    try:
-        per_page = int(url_query["per_page"])
-    except (KeyError, ValueError):
-        per_page = 25
 
     base_query = base_query or {}
 
-    if isinstance(sort, str):
-        sort = {sort: -1 if reverse else 1}
+    client_query = client_query or {}
 
-    db_query = {"$and": [base_query, db_query]}
+    transforms = transforms or []
 
-    found_count = await collection.count_documents(db_query)
+    projection = projection or True
 
-    page_count = int(math.ceil(found_count / per_page))
-
-    total_count = await collection.count_documents(base_query)
+    if isinstance(sort, str):
+        sort = {sort: -1 if reverse else 1}
 
     skip_count = 0
 
-    if found_count and page > 1:
+    if page > 1:
         skip_count = (page - 1) * per_page
 
+    async for paginate_dict in collection.aggregate(
+        [
+            {"$match": base_query},
+            {
+                "$facet": {
+                    "total_count": [
+                        {"$count": "total_count"},
+                    ],
+                    "found_count": [
+                        {"$match": client_query},
+                        {"$count": "found_count"},
+                    ],
+                    "data": [
+                        {
+                            "$match": client_query,
+                        },
+                        {"$sort": sort},
+                        {"$skip": skip_count},
+                        {"$limit": per_page},
+                        *transforms,
```

Comment:
These aren't really transforms are they. They are functions that return lookups for MongoDB aggregations. I think a new name needs to be found.

Reply:
Yeah I will try to think of something else. I was mostly trying to avoid `lookups` as technically these steps could do anything to the data